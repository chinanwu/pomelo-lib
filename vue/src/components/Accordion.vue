<template>
  <div :class="pClass">
    <button
      id="accordionBtn"
      class="AccordionBtn"
      @click="expanded = !expanded"
    >
      <span>{{ label }}</span>
      <span>{{ expanded ? "-" : "+" }}</span>
    </button>
    <transition name="expand" mode="out-in">
      <slot v-if="expanded" name="content"></slot>
    </transition>
  </div>
</template>

<script>
export default {
  name: "Accordion",
  props: {
    label: {
      type: String,
      default: null
    },
    pClass: {
      type: String,
      default: null
    },
    expand: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      expanded: this.expand
    };
  }
};
</script>

<style lang="less" scoped>
.AccordionBtn {
  width: 100%;
  text-align: left;
  padding: 1rem;
  background-color: #efefef;
  border: none;
  display: flex;
  justify-content: space-between;
  cursor: pointer;
  transition: background-color 0.2s;
  border-radius: 0.5rem;
  font-size: 1.4rem;

  &:hover,
  &:focus {
    background-color: #f4f4f4;
  }

  &:active {
    background-color: #efefef;
  }
}

.expand {
  &-enter-active,
  &-leave-active {
    transition: all 0.3s;
    overflow: hidden;
  }
  &-enter-from,
  &-leave-to {
    max-height: 0;
    opacity: 0;
  }
  &-enter-to,
  &-leave-from {
    max-height: 100vh;
    opacity: 1;
  }
}
</style>
