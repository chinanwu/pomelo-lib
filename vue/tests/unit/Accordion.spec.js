import { shallowMount } from "@vue/test-utils";
import Accordion from "@/components/Accordion.vue";

describe("Accordion.vue", () => {
  it("renders Accordion", () => {
    const wrapper = shallowMount(Accordion);
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders expanded accordion with label", () => {
    const wrapper = shallowMount(Accordion, {
      props: {
        label: "Accordion Label",
        expand: true
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("expands on click", () => {
    const wrapper = shallowMount(Accordion);
    wrapper.find("#accordionBtn").trigger("click");
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.vm.expanded).toEqual(true);
  });
});
