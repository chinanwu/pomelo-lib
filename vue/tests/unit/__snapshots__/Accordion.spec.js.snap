// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion.vue expands on click 1`] = `
<div
  class=""
>
  <button
    class="AccordionBtn"
    id="accordionBtn"
  >
    <span />
    <span>
      +
    </span>
  </button>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;

exports[`Accordion.vue renders Accordion 1`] = `
<div
  class=""
>
  <button
    class="AccordionBtn"
    id="accordionBtn"
  >
    <span />
    <span>
      +
    </span>
  </button>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;

exports[`Accordion.vue renders expanded accordion with label 1`] = `
<div
  class=""
>
  <button
    class="AccordionBtn"
    id="accordionBtn"
  >
    <span>
      Accordion Label
    </span>
    <span>
      -
    </span>
  </button>
  <transition-stub>
    
    
  </transition-stub>
</div>
`;
